Mathlib/Algebra/Category/Ring/Topology.lean

```diff
@@ -0,0 +1,176 @@
+lemma isClosedEmbedding_precomp_of_surjective
+    [TopologicalSpace T] [T1Space T] (f : R ⟶ S) (hf : Function.Surjective f) :
+    Topology.IsClosedEmbedding ((f ≫ ·) : (S ⟶ T) → (R ⟶ T)) := by
+  refine ⟨isEmbedding_precomp_of_surjective f hf, ?_⟩
+  have : IsClosed (⋂ i : RingHom.ker f.hom, { f : R ⟶ T | f i = 0 }) :=
+    isClosed_iInter fun x ↦ (isClosed_singleton (x := 0)).preimage (continuous_apply (S := T) x.1)
+  convert this
+  ext x
+  simp only [Set.mem_range, Set.iInf_eq_iInter, Set.mem_iInter, Set.mem_setOf_eq, Subtype.forall,
+    RingHom.mem_ker]
+  constructor
+  · rintro ⟨g, rfl⟩ a ha; simp [ha]
+  · exact fun H ↦ ⟨CommRingCat.ofHom (RingHom.liftOfSurjective f.hom hf ⟨x.hom, H⟩),
+      by ext; simp [RingHom.liftOfRightInverse_comp_apply]⟩
+
+/-- `Hom(S[Xᵢ], R)` is homeomorphic to `Hom(S, R) × Rⁱ`. -/
+@[simps! apply_fst apply_snd symm_apply_hom]
+noncomputable
+def mvPolynomialHomeo (σ : Type v) (R S : CommRingCat.{max u v})
+    [TopologicalSpace R] [IsTopologicalRing R] :
+    (CommRingCat.of (MvPolynomial σ S) ⟶ R) ≃ₜ ((S ⟶ R) × (σ → R)) where
+  toFun f := ⟨CommRingCat.ofHom MvPolynomial.C ≫ f, fun i ↦ f (.X i)⟩
+  invFun fx := CommRingCat.ofHom (MvPolynomial.eval₂Hom fx.1.hom fx.2)
+  left_inv f := by ext <;> simp
+  right_inv fx := by ext <;> simp
+  continuous_toFun := by fun_prop
+  continuous_invFun := by
+    refine continuous_induced_rng.mpr ?_
+    refine continuous_pi fun p ↦ ?_
+    simp only [Function.comp_apply, hom_ofHom, MvPolynomial.coe_eval₂Hom, MvPolynomial.eval₂_eq]
+    fun_prop
+
+variable (R S) in
+lemma isEmbedding_hom [TopologicalSpace S] :
+    IsEmbedding (fun f : R ⟶ S ↦ (f.hom : R → S)) :=
+  ⟨.induced _, fun _ _ e ↦ by ext; rw [e]⟩
+
+open Limits
+
+variable (R S) in
+lemma isClosedEmbedding_hom
+    [TopologicalSpace S] [IsTopologicalRing S] [T1Space S] :
+    IsClosedEmbedding (fun f : R ⟶ S ↦ (f.hom : R → S)) := by
+  let f : CommRingCat.of (MvPolynomial R (⊥_ CommRingCat)) ⟶ R :=
+    CommRingCat.ofHom (MvPolynomial.eval₂Hom (initial.to R).hom id)
+  have : Function.Surjective f := Function.LeftInverse.surjective (g := .X) fun x ↦ by simp [f]
+  convert ((mvPolynomialHomeo R S (.of _)).trans
+    (.uniqueProd (⊥_ CommRingCat ⟶ S) _)).isClosedEmbedding.comp
+    (isClosedEmbedding_precomp_of_surjective f this) using 2 with g
+  ext x
+  simp [f]
+
+instance [TopologicalSpace S] [T2Space S] : T2Space (R ⟶ S) :=
+  (isEmbedding_hom R S).t2Space
+
+instance [TopologicalSpace S] [IsTopologicalRing S] [T1Space S] [CompactSpace S] :
+    CompactSpace (R ⟶ S) :=
+  (isClosedEmbedding_hom R S).compactSpace
+
+open Limits
+
+/-- `Hom(A ⊗[S] B, R)` has the subspace topology from `Hom(A, R) × Hom(B, R)`. -/
+lemma isEmbedding_pushout [TopologicalSpace R] [IsTopologicalRing R]
+    (φ : S ⟶ A) (ψ : S ⟶ B) :
+    IsEmbedding fun f : pushout φ ψ ⟶ R ↦ (pushout.inl φ ψ ≫ f, pushout.inr φ ψ ≫ f) := by
+  let PA := CommRingCat.of (MvPolynomial A S)
+  let PB := CommRingCat.of (MvPolynomial B S)
```

Comment:
This proof looks like it has some interesting ideas in. Why not sketch what's going on in comments? It is long, after all, and after only two lines I'm already convinced that you're doing something interesting.